torchrec/csrc/dynamic_embedding/details/io_registry.h

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+#include <c10/util/flat_hash_map.h>
+#include <dlfcn.h>
+#include <stdint.h>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace tde::details {
+
+struct IOPullParameter {
+  const char* table_name_;
+  uint32_t num_cols_;
+  uint32_t num_global_ids_;
+  const int64_t* col_ids_;
+  const int64_t* global_ids_;
+  uint32_t num_optimizer_stats_;
+  void* on_complete_context_;
+  void (*on_global_id_fetched_)(
+      void* ctx,
+      uint32_t offset,
+      uint32_t optimizer_state,
+      void* data,
+      uint32_t data_len);
+  void (*on_all_fetched_)(void* ctx);
+};
+
+struct IOPushParameter {
+  const char* table_name_;
+  uint32_t num_cols_;
+  uint32_t num_global_ids_;
+  const int64_t* col_ids_;
+  const int64_t* global_ids_;
+  uint32_t num_optimizer_stats_;
+  const uint32_t* optimizer_stats_ids_;
+  // offsets in bytes
+  // data_ptr is 1-D array divided by offsets in bytes.
+  // The offsets are jagged and length of offsets array is length + 1.
+  //
+  // data[global_id * num_cols * num_optimizer_stats_
+  //       + col_id * num_optimizer_stats_ + os_id ]
+  uint32_t num_offsets_;
+  const uint64_t* offsets_;
```

Comment:
Here we were hoping to export an C api, so that user could write the plugin in whatever language they prefer. And that's the reason why we split the `std::span` into a pointer and its size and also split `std::function` into a function pointer and its context.